File: apps/services/notifications/Notifications.Infrastructure/Providers/Adapters/SmtpAdapter.cs
using MailKit.Net.Smtp;
using Microsoft.Extensions.Logging;
using MimeKit;
using Notifications.Application.Interfaces;

namespace Notifications.Infrastructure.Providers.Adapters;

public class SmtpAdapter : IEmailProviderAdapter
{
    public string ProviderType => "smtp";

    private readonly string _host;
    private readonly int _port;
    private readonly bool _useSsl;
    private readonly string _username;
    private readonly string _password;
    private readonly string _fromEmail;
    private readonly string? _fromName;
    private readonly ILogger<SmtpAdapter> _logger;

    public SmtpAdapter(string host, int port, bool useSsl, string username, string password, string fromEmail, string? fromName, ILogger<SmtpAdapter> logger)
    {
        _host = host; _port = port; _useSsl = useSsl;
        _username = username; _password = password;
        _fromEmail = fromEmail; _fromName = fromName;
        _logger = logger;
    }

    public Task<bool> ValidateConfigAsync()
        => Task.FromResult(!string.IsNullOrEmpty(_host) && _port > 0 && !string.IsNullOrEmpty(_username) && !string.IsNullOrEmpty(_password) && !string.IsNullOrEmpty(_fromEmail));

    public async Task<EmailSendResult> SendAsync(EmailSendPayload payload)
    {
        if (!await ValidateConfigAsync())
            return new EmailSendResult { Success = false, Failure = new ProviderFailure { Category = "auth_config_failure", Message = "SMTP config is incomplete", Retryable = false } };

        using var client = new SmtpClient();
        try
        {
            client.Timeout = 10000;
            await client.ConnectAsync(_host, _port, _useSsl ? MailKit.Security.SecureSocketOptions.SslOnConnect : MailKit.Security.SecureSocketOptions.StartTlsWhenAvailable);
            await client.AuthenticateAsync(_username, _password);

            var message = new MimeMessage();
            message.From.Add(new MailboxAddress(_fromName ?? _fromEmail, _fromEmail));
            message.To.Add(MailboxAddress.Parse(payload.To));
            message.Subject = payload.Subject;
            if (payload.ReplyTo != null) message.ReplyTo.Add(MailboxAddress.Parse(payload.ReplyTo));

            var bodyBuilder = new BodyBuilder { TextBody = payload.Body };
            if (payload.Html != null) bodyBuilder.HtmlBody = payload.Html;
            message.Body = bodyBuilder.ToMessageBody();

            var result = await client.SendAsync(message);
            _logger.LogInformation("SMTP: email sent to {To}, messageId={MessageId}", payload.To, result);
            return new EmailSendResult { Success = true, ProviderMessageId = result };
        }
        catch (Exception ex)
        {
            var category = ClassifyError(ex);
            _logger.LogWarning(ex, "SMTP: send failed {Category}", category);
            return new EmailSendResult { Success = false, Failure = new ProviderFailure { Category = category, Message = ex.Message[..Math.Min(ex.Message.Length, 500)], Retryable = category is "retryable_provider_failure" or "provider_unavailable" } };
        }
        finally
        {
            if (client.IsConnected) await client.DisconnectAsync(true);
        }
    }

    public async Task<ProviderHealthResult> HealthCheckAsync()
    {
        if (!await ValidateConfigAsync()) return new ProviderHealthResult { Status = "down" };
        using var client = new SmtpClient();
        var start = DateTime.UtcNow;
        try
        {
            client.Timeout = 5000;
            await client.ConnectAsync(_host, _port, _useSsl ? MailKit.Security.SecureSocketOptions.SslOnConnect : MailKit.Security.SecureSocketOptions.StartTlsWhenAvailable);
            await client.AuthenticateAsync(_username, _password);
            var latencyMs = (int)(DateTime.UtcNow - start).TotalMilliseconds;
            return new ProviderHealthResult { Status = "healthy", LatencyMs = latencyMs };
        }
        catch (Exception ex)
        {
            var latencyMs = (int)(DateTime.UtcNow - start).TotalMilliseconds;
            var category = ClassifyError(ex);
            return new ProviderHealthResult { Status = category == "auth_config_failure" ? "down" : "degraded", LatencyMs = latencyMs };
        }
        finally
        {
            if (client.IsConnected) await client.DisconnectAsync(true);
        }
    }

    private static string ClassifyError(Exception ex)
    {
        var msg = ex.Message.ToLowerInvariant();
        if (msg.Contains("auth") || msg.Contains("credential") || msg.Contains("535") || msg.Contains("534")) return "auth_config_failure";
        if (msg.Contains("timeout") || msg.Contains("refused") || msg.Contains("enotfound")) return "provider_unavailable";
        if (msg.Contains("550") || msg.Contains("invalid") || msg.Contains("recipient")) return "invalid_recipient";
        return "retryable_provider_failure";
    }
}

File: _archived/notifications-nodejs/src/integrations/providers/adapters/smtp.adapter.ts
import * as nodemailer from "nodemailer";
import { EmailProviderAdapter, EmailSendPayload, EmailSendResult } from "../interfaces/email-provider.interface";
import { FailureCategory } from "../../../types";
import { logger } from "../../../shared/logger";

export interface SmtpAdapterConfig {
  host: string;
  port: number;
  secure: boolean;
  username: string;
  password: string;
  fromEmail: string;
  fromName?: string;
}

function classifySmtpError(err: unknown): FailureCategory {
  const msg = String(err).toLowerCase();
  if (msg.includes("auth") || msg.includes("credential") || msg.includes("535") || msg.includes("534")) {
    return "auth_config_failure";
  }
  if (msg.includes("timeout") || msg.includes("econnrefused") || msg.includes("enotfound")) {
    return "provider_unavailable";
  }
  if (msg.includes("550") || msg.includes("invalid") || msg.includes("recipient")) {
    return "invalid_recipient";
  }
  return "retryable_provider_failure";
}

export class SmtpEmailProviderAdapter implements EmailProviderAdapter {
  readonly providerType = "smtp";
  private config: SmtpAdapterConfig;

  constructor(config: SmtpAdapterConfig) {
    this.config = config;
  }

  async validateConfig(): Promise<boolean> {
    const { host, port, username, password, fromEmail } = this.config;
    if (!host || !port || !username || !password || !fromEmail) {
      logger.warn("SMTP config is incomplete", {
        hasHost: !!host,
        hasPort: !!port,
        hasUsername: !!username,
        hasPassword: !!password,
        hasFromEmail: !!fromEmail,
      });
      return false;
    }
    return true;
  }

  async send(payload: EmailSendPayload): Promise<EmailSendResult> {
    const valid = await this.validateConfig();
    if (!valid) {
      return {
        success: false,
        failure: {
          category: "auth_config_failure",
          message: "SMTP config is incomplete — check host, port, username, password, and fromEmail",
          retryable: false,
        },
      };
    }

    const { host, port, secure, username, password, fromEmail, fromName } = this.config;

    const transporter = nodemailer.createTransport({
      host,
      port,
      secure,
      auth: { user: username, pass: password },
      connectionTimeout: 10000,
      greetingTimeout: 5000,
    });

    try {
      const info = await transporter.sendMail({
        from: fromName ? `"${fromName}" <${fromEmail}>` : fromEmail,
        to: payload.to,
        subject: payload.subject,
        text: payload.body,
        html: payload.html ?? payload.body,
        replyTo: payload.replyTo,
      });

      logger.info("SMTP: email sent", { messageId: info.messageId, to: payload.to });
      return { success: true, providerMessageId: info.messageId ?? undefined };
    } catch (err) {
      const message = String(err);
      const category = classifySmtpError(err);
      logger.warn("SMTP: send failed", { error: message, category });

      return {
        success: false,
        failure: {
          category,
          message: message.slice(0, 500),
          retryable: category === "retryable_provider_failure" || category === "provider_unavailable",
        },
      };
    } finally {
      transporter.close();
    }
  }

  async healthCheck(): Promise<{ status: "healthy" | "degraded" | "down"; latencyMs?: number }> {
    const valid = await this.validateConfig();
    if (!valid) return { status: "down" };

    const { host, port, secure, username, password } = this.config;
    const transporter = nodemailer.createTransport({
      host,
      port,
      secure,
      auth: { user: username, pass: password },
      connectionTimeout: 5000,
    });

    const start = Date.now();
    try {
      await transporter.verify();
      const latencyMs = Date.now() - start;
      logger.debug("SMTP health check passed", { host, port, latencyMs });
      return { status: "healthy", latencyMs };
    } catch (err) {
      const latencyMs = Date.now() - start;
      const category = classifySmtpError(err);
      logger.warn("SMTP health check failed", { error: String(err), category });

      if (category === "auth_config_failure") return { status: "down", latencyMs };
      return { status: "degraded", latencyMs };
    } finally {
      transporter.close();
    }
  }
}
